File: web/src/api/system.ts
import { useMutation, useQuery } from '@tanstack/react-query'
import { client } from './client'
import { parseContentDisposition, triggerDownload } from '../utils/download'

export interface UpdateStatus {
  enabled: boolean
  current_commit: string | null
  current_short: string | null
  remote_commit: string | null
  remote_short: string | null
  behind: number
  up_to_date: boolean
  changelog: string[]
  fetch_ok: boolean
  error: string | null
}

export interface ApplyResult {
  ok: boolean
  restarting: boolean
  steps: string[]
  error: string | null
  active_batches: number
}

export interface ToolCheck {
  name: string
  ok: boolean
  detail: string
}

export interface PreflightResult {
  ok: boolean
  tools: ToolCheck[]
  remote_ok: boolean
  remote_detail: string
  tree_clean: boolean
  tree_detail: string
}

/** Read-only readiness check: git/uv/pnpm reachable, remote pullable, clean tree. */
export function usePreflight() {
  return useMutation({
    mutationFn: async () =>
      (await client.get<PreflightResult>('/system/update/preflight')).data,
  })
}

/** Cached local-vs-remote status (no network fetch). Polled by the nav badge. */
export function useUpdateStatus(enabled = true) {
  return useQuery({
    queryKey: ['system', 'update', 'status'],
    queryFn: async () => (await client.get<UpdateStatus>('/system/update/status')).data,
    enabled,
    refetchInterval: 60_000,
  })
}

/** Fetch origin/main and report availability. */
export function useCheckUpdate() {
  return useMutation({
    mutationFn: async () =>
      (await client.post<UpdateStatus>('/system/update/check')).data,
  })
}

/** Pull + rebuild + restart. force=true interrupts in-flight batches. */
export function useApplyUpdate() {
  return useMutation({
    mutationFn: async (force: boolean) =>
      (await client.post<ApplyResult>('/system/update/apply', { force })).data,
  })
}

export interface HealthInfo {
  status: string
  commit: string | null
}

/** Raw /health probe used to detect when a restart has landed. */
export async function probeHealth(): Promise<HealthInfo | null> {
  try {
    const res = await client.get<HealthInfo>('/health', {
      baseURL: '/',
      timeout: 4000,
    })
    return res.data
  } catch {
    return null
  }
}

export interface AppLog {
  path: string
  exists: boolean
  size_bytes: number
  // True when the file has more lines before what's returned — either it
  // was cut off mid-read, or the whole (small) file was read but still has
  // more lines than requested.
  truncated: boolean
  content: string
}

/** Tail the app's own runtime log (Settings.log_file — the whole
 * uvicorn/FastAPI process's stdout+stderr, not just this app's logger). */
export function useAppLog(lines: number, refetchInterval: number | false = false) {
  return useQuery({
    queryKey: ['system', 'log', lines],
    queryFn: async () => (await client.get<AppLog>('/system/log', { params: { lines } })).data,
    refetchInterval,
    placeholderData: (prev) => prev,
  })
}

export async function downloadAppLog(): Promise<void> {
  const response = await client.get('/system/log/download', { responseType: 'blob' })
  const filename =
    parseContentDisposition(response.headers['content-disposition']) ?? 'autoagent.log'
  triggerDownload(response.data as Blob, filename)
}

File: src/autoagent/api/system.py
"""System self-update endpoints: check origin/main and apply + restart.

Gated behind DefaultsConfig.self_update_enabled and admin auth (router-level
require_user). Applying an update interrupts any in-flight batch, so /apply
refuses unless there are none or force=true is passed.
"""

from __future__ import annotations

import asyncio
import logging
from dataclasses import asdict

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import FileResponse
from pydantic import BaseModel

from autoagent.auth.deps import require_user
from autoagent.config.settings import get_settings
from autoagent.models.api import DefaultsConfig
from autoagent.storage.batches import count_active_batches
from autoagent.storage.configs import get_config
from autoagent.system import updater
from autoagent.utils.tail import tail_lines

_log = logging.getLogger(__name__)

router = APIRouter(prefix="/system", tags=["system"], dependencies=[Depends(require_user)])


async def _self_update_enabled() -> bool:
    v = await get_config("defaults")
    cfg = DefaultsConfig.model_validate(v) if v else DefaultsConfig()
    return cfg.self_update_enabled


@router.get("/update/status")
async def update_status() -> dict:
    """Cached local-vs-remote status. Does not hit the network (no fetch)."""
    enabled = await _self_update_enabled()
    status = await asyncio.to_thread(updater.check_for_update, enabled=enabled, do_fetch=False)
    return asdict(status)


@router.get("/update/preflight")
async def update_preflight() -> dict:
    """Diagnose the box (git/uv/pnpm reachable, remote pullable, clean tree).

    Read-only, so it's allowed regardless of the self_update_enabled flag —
    admins can check readiness before turning self-update on.
    """
    result = await asyncio.to_thread(updater.preflight)
    return asdict(result)


@router.post("/update/check")
async def update_check() -> dict:
    """Fetch origin/main and report whether an update is available."""
    enabled = await _self_update_enabled()
    if not enabled:
        raise HTTPException(status_code=403, detail="self-update is disabled")
    status = await asyncio.to_thread(updater.check_for_update, enabled=enabled, do_fetch=True)
    return asdict(status)


class _ApplyRequest(BaseModel):
    force: bool = False


@router.post("/update/apply")
async def update_apply(body: _ApplyRequest) -> dict:
    enabled = await _self_update_enabled()
    if not enabled:
        raise HTTPException(status_code=403, detail="self-update is disabled")

    active = await count_active_batches()
    if active > 0 and not body.force:
        # 409: caller must confirm interrupting in-flight batches with force=true.
        raise HTTPException(
            status_code=409,
            detail={"error": "active_batches", "active_batches": active},
        )

    _log.warning(
        "self-update: applying origin/main (force=%s, active_batches=%d)", body.force, active
    )
    result = await asyncio.to_thread(updater.apply_update)
    payload = asdict(result)
    payload["active_batches"] = active
    if not result.ok:
        raise HTTPException(status_code=500, detail=payload)
    return payload


@router.get("/log")
async def read_app_log(lines: int = Query(default=1000, ge=50, le=10000)) -> dict:
    """Tail the app's own runtime log (Settings.log_file).

    That file is the raw stdout+stderr of the whole uvicorn/FastAPI process
    (app log lines plus uvicorn's own startup/access/error output) —
    written by run.sh's `>>"$LOG_FILE" 2>&1` redirect, not by Python itself.
    It has no rotation, so this always tails rather than risking loading a
    multi-GB file into memory.
    """
    path = get_settings().log_file
    content, truncated = await asyncio.to_thread(tail_lines, path, lines)
    exists = path.exists()
    return {
        "path": str(path.resolve()),
        "exists": exists,
        "size_bytes": path.stat().st_size if exists else 0,
        "truncated": truncated,
        "content": content,
    }


@router.get("/log/download")
async def download_app_log() -> FileResponse:
    path = get_settings().log_file
    if not path.exists():
        raise HTTPException(status_code=404, detail="log file not found")
    return FileResponse(path, filename="autoagent.log", media_type="text/plain")
